File: app/src/androidTest/java/com/paydaytracker/app/BackupTestProvider.java
package com.paydaytracker.app;

import android.database.Cursor;
import android.database.MatrixCursor;
import android.os.Bundle;
import android.os.CancellationSignal;
import android.os.ParcelFileDescriptor;
import android.provider.DocumentsContract;
import android.provider.DocumentsProvider;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

// Standalone test-APK process: use platform Java only, with no target-app dependencies.
public class BackupTestProvider extends DocumentsProvider {
    private boolean failCurrent;
    private File root() { File f=new File(getContext().getFilesDir(),"backup-tests"); f.mkdirs(); return f; }
    private File file(String id) { if (id.equals("root")) return root(); if (!id.startsWith("root/") || id.contains("..")) throw new IllegalArgumentException(); return new File(root(),id.substring(5)); }
    private static final String[] COLUMNS={DocumentsContract.Document.COLUMN_DOCUMENT_ID,DocumentsContract.Document.COLUMN_DISPLAY_NAME,DocumentsContract.Document.COLUMN_MIME_TYPE,DocumentsContract.Document.COLUMN_FLAGS};
    private void row(MatrixCursor c,String id) { File f=file(id); c.addRow(new Object[]{id,id.equals("root")?"Test backups":f.getName(),f.isDirectory()?DocumentsContract.Document.MIME_TYPE_DIR:"application/json",f.isDirectory()?DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE:DocumentsContract.Document.FLAG_SUPPORTS_WRITE|DocumentsContract.Document.FLAG_SUPPORTS_DELETE}); }
    @Override public boolean onCreate(){return true;}
    @Override public Cursor queryRoots(String[] projection){MatrixCursor c=new MatrixCursor(new String[]{DocumentsContract.Root.COLUMN_ROOT_ID,DocumentsContract.Root.COLUMN_DOCUMENT_ID,DocumentsContract.Root.COLUMN_TITLE,DocumentsContract.Root.COLUMN_FLAGS});c.addRow(new Object[]{"root","root","Backup test",DocumentsContract.Root.FLAG_SUPPORTS_CREATE});return c;}
    @Override public Cursor queryDocument(String id,String[] projection){MatrixCursor c=new MatrixCursor(COLUMNS);row(c,id);return c;}
    @Override public Cursor queryChildDocuments(String parent,String[] projection,String order){MatrixCursor c=new MatrixCursor(COLUMNS);File[] files=file(parent).listFiles();if(files!=null)for(File f:files)row(c,"root/"+f.getName());return c;}
    @Override public String createDocument(String parent,String mime,String name)throws FileNotFoundException {if(!parent.equals("root")||name.contains("/"))throw new IllegalArgumentException();try{File f=new File(root(),name);if(!f.createNewFile())throw new IOException();return "root/"+name;}catch(IOException e){throw new FileNotFoundException();}}
    @Override public void deleteDocument(String id)throws FileNotFoundException {if(!file(id).delete())throw new FileNotFoundException();}
    @Override public boolean isChildDocument(String parent,String id){return parent.equals("root")&&id.startsWith("root/");}
    @Override public ParcelFileDescriptor openDocument(String id,String mode,CancellationSignal signal)throws FileNotFoundException {if(failCurrent&&id.endsWith("WageTrack-backup.json")&&mode.contains("w"))throw new FileNotFoundException("Simulated write failure");return ParcelFileDescriptor.open(file(id),ParcelFileDescriptor.parseMode(mode));}
    @Override public Bundle call(String method,String arg,Bundle extras){if(method.equals("fail-current")){failCurrent=true;return new Bundle();}if(method.equals("allow-writes")){failCurrent=false;return new Bundle();}return super.call(method,arg,extras);}
}

File: app/src/androidTest/java/com/paydaytracker/app/BackupTestProvider.kt
package com.paydaytracker.app

import android.app.Activity
import android.content.Intent
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.os.Bundle
import android.os.CancellationSignal
import android.os.ParcelFileDescriptor
import android.provider.DocumentsContract as Docs
import android.provider.DocumentsProvider
import java.io.File

// Instrumentation-only provider: failures affect its private test directory, never user files.
class BackupTestProvider : DocumentsProvider() {
    private var failCurrent=false
    private val root get()=File(requireNotNull(context).filesDir,"backup-tests").apply {mkdirs()}
    override fun onCreate()=true
    private fun file(id:String):File {require(id=="root"||id.startsWith("root/"));val f=if(id=="root")root else File(root,id.removePrefix("root/"));require(f.canonicalPath==root.canonicalPath||f.canonicalPath.startsWith(root.canonicalPath+"/"));return f}
    private fun row(cursor:MatrixCursor,id:String){val f=file(id);cursor.newRow().apply {add(Docs.Document.COLUMN_DOCUMENT_ID,id);add(Docs.Document.COLUMN_DISPLAY_NAME,if(id=="root")"Test backups" else f.name);add(Docs.Document.COLUMN_MIME_TYPE,if(f.isDirectory)Docs.Document.MIME_TYPE_DIR else "application/json");add(Docs.Document.COLUMN_FLAGS,if(f.isDirectory)Docs.Document.FLAG_DIR_SUPPORTS_CREATE else Docs.Document.FLAG_SUPPORTS_WRITE or Docs.Document.FLAG_SUPPORTS_DELETE)}}
    override fun queryRoots(projection:Array<out String>?):Cursor = MatrixCursor(arrayOf(Docs.Root.COLUMN_ROOT_ID,Docs.Root.COLUMN_DOCUMENT_ID,Docs.Root.COLUMN_TITLE,Docs.Root.COLUMN_FLAGS)).apply {addRow(arrayOf("root","root","Backup test",Docs.Root.FLAG_SUPPORTS_CREATE))}
    override fun queryDocument(documentId:String,projection:Array<out String>?):Cursor=MatrixCursor(arrayOf(Docs.Document.COLUMN_DOCUMENT_ID,Docs.Document.COLUMN_DISPLAY_NAME,Docs.Document.COLUMN_MIME_TYPE,Docs.Document.COLUMN_FLAGS)).also {row(it,documentId)}
    override fun queryChildDocuments(parentDocumentId:String,projection:Array<out String>?,sortOrder:String?):Cursor=MatrixCursor(arrayOf(Docs.Document.COLUMN_DOCUMENT_ID,Docs.Document.COLUMN_DISPLAY_NAME,Docs.Document.COLUMN_MIME_TYPE,Docs.Document.COLUMN_FLAGS)).also {c->file(parentDocumentId).listFiles()?.forEach {row(c,"root/"+it.name)}}
    override fun createDocument(parentDocumentId:String,mimeType:String,displayName:String):String {require(parentDocumentId=="root"&&!displayName.contains('/'));val f=File(root,displayName);check(f.createNewFile());return "root/"+f.name}
    override fun deleteDocument(documentId:String){check(file(documentId).delete())}
    override fun isChildDocument(parentDocumentId:String,documentId:String)=parentDocumentId=="root"&&documentId.startsWith("root/")
    override fun openDocument(documentId:String,mode:String,signal:CancellationSignal?):ParcelFileDescriptor {if(failCurrent&&documentId.endsWith(AutoBackup.CURRENT)&&mode.contains('w'))throw java.io.FileNotFoundException("Simulated write failure");return ParcelFileDescriptor.open(file(documentId),ParcelFileDescriptor.parseMode(mode))}
    override fun call(method:String,arg:String?,extras:Bundle?):Bundle? {if(method=="fail-current"){failCurrent=true;return Bundle()};if(method=="allow-writes"){failCurrent=false;return Bundle()};return super.call(method,arg,extras)}
}
class BackupFolderActivity : Activity() {
    override fun onCreate(state:Bundle?) {super.onCreate(state);setResult(RESULT_OK,Intent().setData(Uri.parse("content://com.paydaytracker.backup.tests/tree/root")).addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION or Intent.FLAG_GRANT_PERSISTABLE_URI_PERMISSION or Intent.FLAG_GRANT_PREFIX_URI_PERMISSION));finish()}
}
